File: python/linkedlist/singly_linked_list.py

# 单链表实现


# 节点定义
class ListNode:

    def __init__(self, val):
        self.next = None
        self.val = val


class LinkedList:

    def __init__(self):
        self.head = None

    # 根据值查找节点
    def find(self, val):
        p = self.head
        while p and p.val != val:
            p = p.next
        return p

    # 添加节点至链表末尾
    def add(self, val) -> ListNode:
        node = ListNode(val)
        if not self.head:
            self.head = node
        else:
            p = self.head
            while p.next:
                p = p.next
            p.next = node
        return node

    # 在某个节点之前添加
    def addBefore(self, p: ListNode, val):
        if not p:
            return
        node = ListNode(val)
        # 在头节点之前插入
        if p is self.head:
            node.next = self.head
            self.head = node
        else:
            prev = self.head
            while prev and prev.next != p:
                prev = prev.next
            if prev:
                node.next = p
                prev.next = node

    # 在某个节点之后插入
    def addAfter(self, p: ListNode, val):
        if not p:
            return
        node = ListNode(val)
        node.next = p.next
        p.next = node

    # 删除节点
    def delete(self, node: ListNode) -> bool:
        p = self.head
        while p and p.next != node:
            p = p.next
        if not p:
            return False

        p.next = p.next.next
        return True

    # 打印所有元素值
    def print(self):
        if not self.head:
            return
        p = self.head
        while p:
            print(p.val, end=' ')
            p = p.next
        print()


if __name__ == '__main__':
    lis = LinkedList()
    lis.add(1)
    n1 = lis.add(3)
    n2 = lis.add(5)

    lis.print()

    lis.delete(n2)
    lis.print()

    lis.addBefore(n1, 10)
    lis.print()

File: go/linkedlist/singly_linkedlist/singly_linked_list.go
package singly_linkedlist

import "fmt"

//单链表的实现

//节点定义
type ListNode struct {
	Val  interface{}
	Next *ListNode
}

//创建节点
func newNode(val interface{}) *ListNode {
	return &ListNode{val, nil}
}

//单链表结构
type LinkedList struct {
	head *ListNode
	size int
}

func New() *LinkedList {
	return &LinkedList{nil, 0}
}

//增加数据至链表末尾
func (lis *LinkedList) PushBack(val interface{}) *ListNode {
	p := lis.head

	node := newNode(val)
	if p == nil {
		lis.head = node
	} else {
		for p.Next != nil {
			p = p.Next
		}
		p.Next = node
	}

	lis.size++
	return node
}

//在某个节点之前插入
func (lis *LinkedList) PushBefore(p *ListNode, val interface{}) *ListNode {
	if p == nil || lis.head == nil {
		return nil
	}

	node := newNode(val)

	if p == lis.head {
		node.Next = lis.head
		lis.head = node
	} else {
		prev := lis.head
		for ; prev.Next != p; prev = prev.Next {
		}

		node.Next = p
		prev.Next = node
	}
	lis.size++
	return node
}

//在某个节点之后插入
func (lis *LinkedList) PushAfter(p *ListNode, val interface{}) *ListNode {
	if p == nil {
		return nil
	}

	node := newNode(val)
	node.Next = p.Next
	p.Next = node
	lis.size++
	return node
}

//根据值查找链表节点
func (lis *LinkedList) Find(val interface{}) *ListNode {
	p := lis.head
	for p != nil && p.Val != val {
		p = p.Next
	}

	return p
}

//删除节点
func (lis *LinkedList) Delete(p *ListNode) {
	if p == nil {
		return
	}

	if p == lis.head {
		lis.head = lis.head.Next
		lis.size--
	} else {
		prev := lis.head
		for prev != nil && prev.Next != p {
			prev = prev.Next
		}

		if prev != nil {
			prev.Next = p.Next
			lis.size--
		}
	}
}

//删除值为目标的节点
func (lis *LinkedList) DeleteVal(val interface{}) {
	lis.Delete(lis.Find(val))
}

func (lis *LinkedList) Size() int {
	return lis.size
}

//打印链表中所有数据
func (lis *LinkedList) PrintData() {
	for p := lis.head; p != nil; p = p.Next {
		fmt.Print(p.Val, " ")
	}
	fmt.Println()
}
